src/adapter/src/coord/sequencer/inner/create_continual_task.rs

```diff
@@ -24,6 +43,160 @@ impl Coordinator {
         plan: plan::CreateContinualTaskPlan,
         resolved_ids: ResolvedIds,
     ) -> Result<ExecuteResponse, AdapterError> {
-        todo!("WIP {:?}", (session, plan, resolved_ids));
+        let plan::CreateContinualTaskPlan {
+            name,
+            desc,
+            input_id,
+            continual_task:
+                plan::MaterializedView {
+                    create_sql,
+                    cluster_id,
+                    mut expr,
+                    column_names,
+                    non_null_assertions,
+                    compaction_window: _,
+                    refresh_schedule,
+                    as_of: _,
+                },
+        } = plan;
+        // WIP feature flag
+
+        // Collect optimizer parameters.
+        let compute_instance = self
+            .instance_snapshot(cluster_id)
+            .expect("compute instance does not exist");
+
+        let debug_name = self.catalog().resolve_full_name(&name, None).to_string();
+        let optimizer_config = optimize::OptimizerConfig::from(self.catalog().system_config())
+            .override_from(&self.catalog.get_cluster(cluster_id).config.features());
+
+        let view_id = self.allocate_transient_id();
+        let catalog_mut = self.catalog_mut();
+        let sink_id = catalog_mut.allocate_user_id().await?;
+
+        // Put a placeholder in the catalog so the optimizer can find something
+        // for the sink_id.
+        let fake_entry = CatalogEntry {
+            item: CatalogItem::Table(Table {
+                create_sql: Some(create_sql.clone()),
+                desc: desc.clone(),
+                conn_id: None,
+                resolved_ids: resolved_ids.clone(),
+                custom_logical_compaction_window: None,
+                is_retained_metrics_object: false,
+                data_source: TableDataSource::TableWrites {
+                    defaults: Vec::new(),
+                },
+            }),
+            referenced_by: Vec::new(),
+            used_by: Vec::new(),
+            id: sink_id,
+            oid: 0,
+            name: name.clone(),
+            owner_id: *session.current_role_id(),
+            privileges: PrivilegeMap::new(),
+        };
+        catalog_mut.hack_add_ct(sink_id, fake_entry);
```

Comment:
> If that's true, this would be an easy-ish refactor? (Definitely something for a followup, even if it's that easy.)

Might be true, I also don't have a lot more context. In my mind the optimizer needs the catalog only to know the types of the objects a dataflow is interacting with, and which indexes are available. I'd say it's definitely worth to take a shot at the refactor and see if we encounter any blockers.

> On the other hand, the optimization part of CTs is basically identical to MVs and I worry that going down this road will prevent us from reusing a bunch of existing infra, in particular stuff like EXPLAIN

Hm, I didn't consider the EXPLAIN stuff. I wonder if it's not possible to make that generic over the concrete optimizer type. We already have the `Optimize` trait and at a glance that should be sufficient for this purpose.

Reply:
> I might have missed something but I just quickly checked and it seems like Optimizer just gets the `CatalogState` from this to hand to `DataflowBuilder` and then `DataflowBuilder` only uses `get_entry`, `resolve_full_name`, and `get_indexes_on` from the `CatalogState`. If that's true, this would be an easy-ish refactor? (Definitely something for a followup, even if it's that easy.)

Wow yeah maybe it really is that easy? https://github.com/MaterializeInc/materialize/pull/29598